design-proposals/vm-memory-dump.md

```diff
@@ -0,0 +1,147 @@
+It will be possible to track the last memory dump in the VM status with a new memory dump status that will be added there.
+It will be possible to unbound the PVC and then do a memory dump to another PVC that will be bound instead.
+
+## API
+
+### Get memory dump
+#### trigger memory dump
+Run a virtctl command to get a memory dump. This new API will look as follows:
+
+`$ virtctl memory-dump my-vm --volume-name=memory-pvc`
+
+In this case `memory-pvc` should already exist and be of a size big enough to contain the memory dump. A check will be made to make sure.
+It will be possible to ask for a PVC to be created with `--create` flag. In such case the required size will be calculated.
+
+##### The process
+The trigger of memory dump will call a VM subresource endpoint which will look like this:
+`/apis/subresources.kubevirt.io/v1alpha3/namespaces/<vm-namespace>/virtualmachines/<vm-name>/memorydump`
+
+The VM subresource endpoint will recieve a rest request containing the pvc name the user want to dump the memory to.
+It will patch the vm status with a memoryDumpStatus stating the volume-name and Phase `Binding`.
+The memoryDumpStatus will look something like this:
+
+`
+// MemoryDumpStatus represent the memory dump request status and info
+type MemoryDumpStatus struct {
+	// VolumeName is the name of the pvc that will contain the memory dump
+	VolumeName string `json:"volumeName"`
+	// Phase represents the memory dump phase
+	Phase MemoryDumpPhase `json:"phase,omitempty"`
+	// TimeStamp represents the time the memory dump was completed
+	TimeStamp *metav1.Time `json:"timestamp,omitempty"`
+}
+
+type MemoryDumpPhase string
+
+const (
+	// The memorydump is during pvc binding
+	MemoryDumpBinding MemoryDumpPhase = "Binding"
+	// The memorydump is in progress
+	MemoryDumpInProgress MemoryDumpPhase = "InProgress"
+	// The memorydump is completed
+	MemoryDumpCompleted MemoryDumpPhase = "Completed"
+	// The memorydump is being unbound
+	MemoryDumpUnBinding MemoryDumpPhase = "Unbinding"
+)
+`
+
+In the virt controller once getting the update of the memoryDump in state of binding it will bind the pvc to the vm by updating the vm volumes with the memoryDump pvc.
+After that the status of the memoryDump will be updated to `InProgress` which will be cause the volume to be mounted to the virt-launcher, same as done in the hotplug process, other then the end of the process that instead of attaching the volume to the VMI it will trigger the guest memory dump `virDomainCoreDump` command with `VIR_DUMP_MEMORY_ONLY` flag to be executed by virt-launcher.
+(Check out [virDomainCoreDump](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainCoreDump) for reference)
+After the dump is complete the timestamp and status of the memory dump will be updated and it will unmount the pvc from virt-launcher.
```

Comment:
The unmount has to happen in the virt-handler pod, so we need a mechanism to communicate that we are complete from the virt-launcher to the virt-controller, so the virt-controller can then update the status which in turn triggers the virt-handler to unmount the volume.

Reply:
You can look at the PR I created for the memory dump, the action in of the memorydump triggers an event on the domain which updates the timestamp and completion of the memorydump. in a future PR this update of state in the volume status in the vmi will triger the update of that status on the vm status which will then remove the memorydump from the vmi which will cause the unmount